File: Android/service/src/main/java/com/scurab/android/anuitor/extract/view/ReflectionExtractor.java
package com.scurab.android.anuitor.extract.view;

import android.util.Log;

import com.scurab.android.anuitor.extract2.BaseExtractor;
import com.scurab.android.anuitor.reflect.Reflector;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * User: jbruchanov
 * Date: 12/05/2014
 * Time: 14:20
 */
public class ReflectionExtractor extends BaseExtractor {
    private boolean mUseFields;

    /**
     * Fill this with regexp patterns to ignore methods
     */
    public static Pattern[] IGNORE_PATTERNS = new Pattern[]{
            Pattern.compile("add.*", 0),
            Pattern.compile("call.*", 0),
            Pattern.compile("clear.*", 0),
            Pattern.compile("create.*", 0),
            Pattern.compile("dispatch.*", 0),
            Pattern.compile("exec.*", 0),
            Pattern.compile("find.*", 0),
            Pattern.compile("gen.*", 0),
            Pattern.compile("mutate.*", 0),
            Pattern.compile("on.*", 0),
            Pattern.compile("perform.*", 0),
            Pattern.compile("pop.*", 0),
            Pattern.compile("post.*", 0),
            Pattern.compile("request.*", 0),
            Pattern.compile("resolve.*", 0),
            Pattern.compile("select.*", 0),
            Pattern.compile("show.*", 0),
            Pattern.compile("will.*", 0)
    };

    public ReflectionExtractor() {
        this(false);
    }

    public ReflectionExtractor(boolean useFields) {
        mUseFields = useFields;
    }

    @NotNull
    @Override
    protected Map<String, Object> onFillValues(@NotNull Object item, @NotNull Map<String, Object> data, @Nullable Map<String, Object> contextData) {
        return fillValues(item, data, contextData, new HashSet<>(), 0);
    }

    private Map<String, Object> fillValues(Object o, Map<String, Object> data, Map<String, Object> contextData, Set<Object> cycleHandler, int depth) {
        data.put("Type", String.valueOf(o.getClass().getName()));
        data.put("ToString", String.valueOf(o));
        Class clz = o.getClass();
        List<Method> methods = getAllMethods(new ArrayList<Method>(), clz);
        for (Method method : methods) {
            String name = method.getName();
            if (ignoreMethod(name)) {
                continue;
            }

            Class<?> returnType = method.getReturnType();
            if (method.getParameterTypes().length == 0
                    && returnType.isPrimitive()
                    && !returnType.equals(Void.TYPE)) {

                method.setAccessible(true);
                try {
                    Object value = method.invoke(o, (Object[])null);
                    data.put(name, value);
                } catch (Exception e) {
                    Log.e("Extractor", String.format("Name:%s Exception:%s", name, e.getClass().getSimpleName()));
                }
            }
        }
        if (mUseFields && depth < 2) {
            final List<Field> fields = getAllFields(new ArrayList<Field>(), o.getClass());
            for (Field f : fields) {
                String name = f.getName();
                if (Modifier.isStatic(f.getModifiers()) || name.startsWith("shadow$")) {
                    continue;
                }
                try {
                    f.setAccessible(true);
                    Object value = f.get(o);
                    if (value != null && !cycleHandler.contains(value)) {
                        cycleHandler.add(value);
                        final Class<?> aClass = Reflector.fixAutoboxing(value.getClass());
                        if (aClass.isPrimitive() || aClass == String.class) {
                            data.put(name, value);
                        } else if (value instanceof Object[]) {
                            List<Object> result = new ArrayList<>();
                            for (Object v : (Object[]) value) {
                                if (v != null) {
                                    final Class<?> vClass = Reflector.fixAutoboxing(v.getClass());
                                    if (vClass.isPrimitive() || vClass == String.class) {
                                        result.add(v);
                                    } else {
                                        result.add(fillValues(v, new HashMap<String, Object>(), data, cycleHandler, depth + 1));
                                    }
                                }
                            }
                            data.put(name, result);
                        } else {
                            data.put(name, fillValues(value, new HashMap<String, Object>(), data, cycleHandler, depth + 1));
                        }
                    }
                } catch (Throwable t) {
                    t.printStackTrace();
                }
            }
        }

        return data;
    }

    public static boolean ignoreMethod(String methodName) {
        for (Pattern s : IGNORE_PATTERNS) {
            if (s.matcher(methodName).matches()) {
                return true;
            }
        }
        return false;
    }

    private static List<Field> getAllFields(List<Field> fields, Class<?> type) {
        fields.addAll(Arrays.asList(type.getDeclaredFields()));
        if (type.getSuperclass() != null) {
            fields = getAllFields(fields, type.getSuperclass());
        }
        return fields;
    }

    private static List<Method> getAllMethods(List<Method> methods, Class<?> type) {
        methods.addAll(Arrays.asList(type.getDeclaredMethods()));
        if (type.getSuperclass() != null) {
            methods = getAllMethods(methods, type.getSuperclass());
        }
        return methods;
    }
}

File: Android/service/src/main/java/com/scurab/android/anuitor/extract2/ReflectionExtractor.kt
package com.scurab.android.anuitor.extract2

import android.os.Handler
import android.os.MessageQueue
import android.util.Log
import java.lang.reflect.Modifier
import java.util.regex.Pattern

private const val MAX_DEPTH = 2

class ReflectionExtractor(private val useFields: Boolean = false, private val maxDepth: Int = MAX_DEPTH) : BaseExtractor() {
    companion object {
        /**
         * Fill this with regexp patterns to ignore methods
         */
        private val IGNORE_PATTERNS = arrayOf(
                Pattern.compile("add.*", 0),
                Pattern.compile("as.*", 0),
                Pattern.compile("begin.*", 0),
                Pattern.compile("call.*", 0),
                Pattern.compile("clear.*", 0),
                Pattern.compile("clone.*", 0),
                Pattern.compile("commit.*", 0),
                Pattern.compile("create.*", 0),
                Pattern.compile("dispatch.*", 0),
                Pattern.compile("exec.*", 0),
                Pattern.compile("find.*", 0),
                Pattern.compile("gen.*", 0),
                Pattern.compile("lock.*", 0),
                Pattern.compile("mutate.*", 0),
                Pattern.compile("on.*", 0),
                Pattern.compile("open.*", 0),
                Pattern.compile("obtain.*", 0),
                Pattern.compile("perform.*", 0),
                Pattern.compile("pop.*", 0),
                Pattern.compile("post.*", 0),
                Pattern.compile("request.*", 0),
                Pattern.compile("resolve.*", 0),
                Pattern.compile("save.*", 0),
                Pattern.compile("select.*", 0),
                Pattern.compile("show.*", 0),
                Pattern.compile("will.*", 0))

        private val IGNORE_CLASSES = setOf<Class<*>>(MessageQueue::class.java, Handler::class.java)
    }

    override fun onFillValues(item: Any, data: MutableMap<String, Any>, contextData: MutableMap<String, Any>?, depth: Int): MutableMap<String, Any> {
        super.onFillValues(item, data, contextData, depth)
        return this.fillValues(item, data, contextData ?: mutableMapOf(), mutableSetOf(), depth)
    }

    private fun fillValues(item: Any, data: MutableMap<String, Any>, contextData: MutableMap<String, Any>?, cycleHandler: MutableSet<Any>, depth: Int): MutableMap<String, Any> {
        if (depth >= maxDepth) {
//            data["depth"] = "max:$depth reached"
            return data
        }
        item.allMethods()
                .filter { !it.name.shouldIgnore() }
                .filter { it.parameterTypes.isEmpty() /*&& it.returnType.isPrimitive*/ && it.returnType != Void.TYPE }
                .forEach { m ->
                    try {
                        Log.d("ReflectionExtractor", "Method:${m.name} Object:$item")
                        m.isAccessible = true
                        m.invoke(item)?.let { v ->
                            storeItem(m.name, v, data, cycleHandler, depth)
                        }
                    } catch (e: Throwable) {
                        Log.e("ReflectionExtractor", "Name:${m.name} Object:$item Exception:${e.javaClass.simpleName}")
                    }
                }

        if (useFields) {
            item.allFields()
                    .filter { !(Modifier.isStatic(it.modifiers) || it.name.startsWith("shadow$")) }
                    .forEach { f ->
                        try {
                            Log.d("ReflectionExtractor", "Name:${f.name} Object:$item")
                            f.isAccessible = true
                            f.get(item)?.let { v ->
                                storeItem(f.name, v, data, cycleHandler, depth)
                            }
                        } catch (e: Throwable) {
                            Log.e("ReflectionExtractor", "Name:${f.name} Object:$item Exception:${e.javaClass.simpleName}")
                        }
                    }
        }
        return data
    }

    private fun storeItem(name: String, v: Any, data: MutableMap<String, Any>, cycleHandler: MutableSet<Any>, depth: Int) {
        if (depth >= maxDepth) {
            return
        }
        if (IGNORE_CLASSES.contains(v.javaClass)) {
            return
        }
        if (v.javaClass.isPrimitiveType()) {
            data[name] = v
        } else if (!cycleHandler.contains(v)) {
            cycleHandler.add(v)
            data[name] = v.asCollection()
                    ?.let { collection ->
                        var result: Any? = null
                        if (depth < maxDepth) {
                            result = collection.map { item ->
                                item?.let {
                                    if (it.javaClass.isPrimitiveType()) {
                                        it
                                    } else {
                                        it.toString()
                                    }
                                }
                            }
                        }
                        result
                    }
                    //no specific extractors, that could potentially create circular reference
                    ?: this.fillValues(v, mutableMapOf(), data, cycleHandler, depth + 1)
        } else {
            data[name] = v.toString()
        }
    }

    private fun String.shouldIgnore(): Boolean {
        IGNORE_PATTERNS.forEach { p ->
            if (p.matcher(this).matches()) {
                return true
            }
        }
        return false
    }

    private fun Class<*>.isPrimitiveType(): Boolean {
        return this.isPrimitive or when (this) {
            Int::class.java,
            Int::class.javaPrimitiveType,
            Int::class.javaObjectType,
            Boolean::class.java,
            Boolean::class.javaPrimitiveType,
            Boolean::class.javaObjectType,
            Short::class.java,
            Short::class.javaPrimitiveType,
            Short::class.javaObjectType,
            Char::class.java,
            Char::class.javaPrimitiveType,
            Char::class.javaObjectType,
            Byte::class.java,
            Byte::class.javaPrimitiveType,
            Byte::class.javaObjectType,
            Long::class.java,
            Long::class.javaPrimitiveType,
            Long::class.javaObjectType,
            Double::class.java,
            Double::class.javaPrimitiveType,
            Double::class.javaObjectType,
            Float::class.java,
            Float::class.javaPrimitiveType,
            Float::class.javaObjectType,
            String::class.java,
            String::class.javaPrimitiveType,
            String::class.javaObjectType -> true
            else -> false
        }
    }

    private fun Any.asCollection() : Collection<*>? {
        return when {
            this is Array<*> -> return this.toList()
            this is IntArray -> return this.toList()
            this is LongArray -> return this.toList()
            this is ByteArray -> return this.toList()
            this is ShortArray -> return this.toList()
            this is BooleanArray -> return this.toList()
            this is CharArray -> return this.toList()
            this is Collection<*> -> return this
            else -> null
        }
    }
}